### PR TITLE
feat: adding cert template for sub ca creation

### PR DESCRIPTION
**Pull request by sepulworld**
Signed-off-by: Zane Williamson <zwilliamson@flexport.com>

* Adds support to request CA certs from aws-privateca-issuer.

### CODE REVIEW

**Comment by bmsiegel** [comment]
Thanks for submitting this pull request!  Pull request reviews and check-ins are on-hold while we update testing. We will review pull requests in the order we received them once the update is complete. Please check back to view the status of the update and your pull request.

**Comment by anbaig** [comment]
Thanks for taking the time to cut this PR! Apologies for the delay on a review, just wanted to make sure I understood the implications of this PR before approving for merge. My largest concern here is that this is a one way door in terms of the use of the `isCA` flag. If this PR got merged as is, we would be forever committed to supporting the `acm-pca:::template/SubordinateCACertificate_PathLen0/V1` template when that flag was toggled. It doesn’t leave room for the use of other CA templates, such as `arn:aws:acm-pca:::template/SubordinateCACertificate_PathLen1/V1` or the other path length/passthrough templates. I’d be more willing to approve this for merge if we could address this issue. Some ideas I had:

- Perhaps we could make aws-privateca-issuers configurable with a flag TemplateArn that if set would make it so that any certificate issued off that issuer uses that template?
- Alternatively, we would add a flag that is called SubCAPathLen, which is required to be set if the isCA flag is toggled?

 Thoughts?

**Comment by sepulworld** [comment]
This did come to mind when putting this pull request in. However, there are presently no configuration interfaces for this plugin ("isCA" is part of the cert-manager object that is passed to plugin).  This means we will have to introduce something new to the plugins design that is a bit outside of intended scope. (which I am fine with!). 

I think option 2 sounds good. Ill try to find time to add the api interface and update this

**Comment by sepulworld** [comment]
Can we add a new spec input without having to go to cert-manager to update interface there?

for example adding path length as an input...

```
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: pca-issuer-rsa-ca-cert
  namespace: default
spec:
  isCA: true
  subCAPathLen: 1
---snip---
```

Let me try it out and see if it works for us

**Comment by anbaig** [comment]
> Can we add a new spec input without having to go to cert-manager to update interface there?
> 
> for example adding path length as an input...
> 
> ```
> apiVersion: cert-manager.io/v1
> kind: Certificate
> metadata:
>   name: pca-issuer-rsa-ca-cert
>   namespace: default
> spec:
>   isCA: true
>   subCAPathLen: 1
> ---snip---
> ```
> 
> Let me try it out and see if it works for us

I don't think you can update the Certificate Kind without cutting a PR to Cert-Manager. The creation of the Certificate will kick off the Cert-Manager controller which will try to generate a CSR but will probably run into issues when it sees "subCAPathLen" and doesn't understand what that is.

**Comment by sepulworld** [comment]
> without

Ok, any suggestions on how we should introduce variable inputs for this issuer?

**Comment by sepulworld** [comment]
> > without
> 
> Ok, any suggestions on how we should introduce variable inputs for this issuer?

I just reread your initial suggestion. By adding a flag you are referring to a flag for the process (https://github.com/cert-manager/aws-privateca-issuer/blob/master/main.go#L54) where we can define the pathLength. Let me take a look into this.

**Comment by anbaig** [comment]
When I meant adding a flag, I meant introducing a new attribute into the AWSPCAIssuer/ClusterIssuer kind. Sort of like how it was done in this PR: https://github.com/cert-manager/aws-privateca-issuer/pull/42/files.

**Comment by diranged** [comment]
So - just to be clear the concern here is that the current code (ultimately a VERY simple change) is not flexible enough to allow creating a SubCA that can then create more SubCAs under it, right? This is something that AWS barely supports on their own right now, and in fact prohibits you from doing across accounts through the Resource Access Manager service anyways.

Can we get this merged as-is for now? If you _really_ want it to be configurable, make it a flag that is set on the aws-privateca pod at runtime.. sort of a cluster-level setting?

@anbaig Do you have a built-docker-image of this patch that we can test out?

**Comment by diranged** [comment]
@anbaig So I was able to build the customized build with https://github.com/diranged/aws-privateca-issuer/commit/09c20605230f5d51e4f414321625172c9da7185e and prove that this works, and works well. I strongly suggest finishing the PR (removing the [check](https://github.com/diranged/aws-privateca-issuer/commit/09c20605230f5d51e4f414321625172c9da7185e#diff-0023dd1178fb9031823ec9ecc0455654147cb6672a3170aa53c73dcc663e0054L135)) and getting this out.. it's great.

**Comment by anbaig** [comment]
@diranged You make a fair point. We could always do that configuration for the path length after the fact. So I went ahead and created a feature request #98 for adding the configuration for the subCA path length. We can go ahead and merge this in.
@sepulworld if you have done any work on making the path length configurable, you can link that PR in #98. Thank you for putting this PR together! Much appreciated :)

File: pkg/aws/pca.go
Comment (diranged) [inline pkg/aws/pca.go:184]: Quick reminder - you also need to remove the check like I did at https://github.com/diranged/aws-privateca-issuer/commit/09c20605230f5d51e4f414321625172c9da7185e#diff-0023dd1178fb9031823ec9ecc0455654147cb6672a3170aa53c73dcc663e0054

**Comment by anbaig** [comment]
Hey all, so I chatted with the team about this and we want to make sure that whatever we set as this default is sensible and supports a variety usecases and ensure there aren't any accidental one way doors we are putting ourselves through. We plan to host a public forum to discuss these issues and get various opinions before making a decision. More details to come about joining the public forum discussion.

**Comment by anbaig** [comment]
We will have a meeting via Amazon Chime to discuss this PR further. You can join via the browser at: 
https://chime.aws/4966718834

or if you have Amazon Chime installed you can join via meeting ID: 4966718834

The call will take place 1-2PM EST, Dec 10. 2021.

**Comment by divyansh-gupta** [comment]
The meeting took place. Here are meeting notes:

Attendees: Divyansh Gupta, Akbar Baig, Brandy Kinlaw, Dan Choi, Nishanth Solomon, Brady Siegal
Date: Dec 10, 2021

Akbar: shared his screen and the Github Issue was discussed. What are the right defaults for a CA template?
Divy: discussion on whether a default makes sense - will that create a 1-way door in the future, in case we want to add a templateArn field?

Ideas:
- Creating a PathLen field to let users choose their own PathLen.
- Creating a TemplateArn field on the Issuer.
- What should the default CA certificate be?
- `BlankSubordinateCACertificate_PathLen0_APICSRPassthrough` as a reasonable default.

Decisions made:
- We want to gather user data on creating a TemplateArn field - are users ok with creating a new issuer per template? how many templates are customers using, usually?
- We will merge this PR in with `BlankSubordinateCACertificate_PathLen0_APICSRPassthrough` as a reasonable default and will add integration tests.

**Comment by diranged** [comment]
> * We will merge this PR in with `BlankSubordinateCACertificate_PathLen0_APICSRPassthrough` as a reasonable default and will add integration tests.

Quick note.. if you do this, we'll have to continue to run our own image. Amazon's Cross-Account sharing system (RAM - Resource Access Manager) specifically articulates the exact TemplateARN that they support:

image

While the differences between [`BlankSubordinateCACertificate_PathLen0_APICSRPassthrough`](https://docs.aws.amazon.com/acm-pca/latest/userguide/UsingTemplates.html#BlankSubordinateCACertificate_PathLen0_APICSRPassthrough) and [`SubordinateCACertificate_PathLen0 `](https://docs.aws.amazon.com/acm-pca/latest/userguide/UsingTemplates.html#SubordinateCACertificate_PathLen0-V1)’ seem minor, Amazon is explicitly forcing our hand to use the `SubordinateCACertificate_PathLen0` template.

**Comment by anbaig** [comment]
@diranged thanks for bringing up the issue about the sub CA template that came out of the discussion not being supported by cross account. I went ahead and made the default the Path Len zero template that is not the blank or APICSR passthrough variant. I also went ahead and added relevant unit tests and integ test to the branch @sepulworld!

**Comment by diranged** [comment]
> @diranged thanks for bringing up the issue about the sub CA template that came out of the discussion not being supported by cross account. I went ahead and made the default the Path Len zero template that is not the blank or APICSR passthrough variant. I also went ahead and added relevant unit tests and integ test to the branch @sepulworld!

Thank you!

**Review by divyansh-gupta — APPROVED** [review]

**Pull request merged**
